src/tracing/service/tracing_service_impl.cc

```diff
@@ -276,6 +277,39 @@ base::ScopedFile CreateTraceFile(const std::string& path, bool overwrite) {
   return fd;
 }
 
+// TODO(ktimofeev): implement for other OSes
+// TODO(ktimofeev): add fallback to read/write loop if sendfile doesn't work
+// See chromium implementation for details:
+// https://source.chromium.org/chromium/chromium/src/+/main:base/files/file_util.cc;l=190;drc=134eb3609e728c12c60a92656f8bca1c888947c1
+base::Status CopyFile(base::PlatformHandle from_fd,
+                      base::PlatformHandle to_fd) {
+#if PERFETTO_BUILDFLAG(PERFETTO_OS_ANDROID) || \
+    PERFETTO_BUILDFLAG(PERFETTO_OS_LINUX)
+  auto maybe_file_size = base::GetFileSize(from_fd);
+  if (!maybe_file_size)
+    return base::ErrStatus("Can't get size of 'from_fd' file.");
+  uint64_t file_size = maybe_file_size.value();
+
+  size_t copied = 0;
+  ssize_t res = 0;
+  do {
+    // Don't specify an offset and the kernel will begin reading/writing to the
+    // current file offsets.
+    res = PERFETTO_EINTR(
+        sendfile(to_fd, from_fd, /*offset=*/nullptr,
```

Comment:
if you use sendfile, you should:

1. store the original seek position
2. lseek(0, SEEK_SET)
3. do the sendfile
4. restore the seek position with lseek(prev, SEEK_SET)


but also I wouln't use sendfile here, because the signature is different on MacOs, and you'd have to work with both. Also windows.

Honestly rather than all this drama, I'd just implement it in file_utils.h/cc and for now just use pread/pwrite

Also I think (i am not sure tbh) you might have to call FlushFile on the source before reading it back. i vaguely remember some problems in som etests that were doing somethign similar before